src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -469,18 +468,32 @@ private static bool CanUseScatterGatherWindowsAPIs(SafeFileHandle handle)
                         return false;
                     }
 
-                    MemoryHandle handle = handlesToDispose[i] = handler.Pin(in buffer);
-                    long ptr = segments[i] = (long)handle.Pointer;
+                    MemoryHandle handle = handler.Pin(in buffer);
+                    long ptr = (long)handle.Pointer;
                     if ((ptr & alignedAtPageSizeMask) != 0)
                     {
+                        handle.Dispose();
                         return false;
                     }
+
+                    // We avoid allocations if there are no
+                    // buffers or the first one is unacceptable.
+                    (handlesToDispose ??= new MemoryHandle[buffersCount])[i] = handle;
+                    if (segments == null)
+                    {
+                        // "The array must contain enough elements to store nNumberOfBytesToWrite
+                        // bytes of data, and one element for the terminating NULL."
+                        segments = (long*)NativeMemory.Alloc((nuint)(buffersCount + 1), sizeof(long));
+                    }
+                    segments[i] = ptr;
                 }
 
+                Debug.Assert(segments != null);
```

Comment:
I might be wrong, but it most likely be `false` for an empty list of buffers. Could you please add a test for this scenario?

Reply:
Nice catch, indeed it would be `null`. I fixed it by setting the last element to zero the soonest the memory is allocated, but added a test nevertheless.